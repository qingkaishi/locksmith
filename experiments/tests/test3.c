#include <pthread.h>
#include <stdio.h>

int f(pthread_mutex_t *l) {
  int k, *j;
  
  pthread_mutex_init(l, NULL);
  //pthread_mutex_lock(l);

  j = &k;
	*j = 3;
  if(pthread_mutex_trylock(l)) {
    *j = 5;
		pthread_mutex_unlock(l);
  }
  return 2;
}

void * dostuff(void * a) {
  pthread_mutex_t t;
  printf("foo");
	pthread_mutex_lock(&t);
  *(int *)a = 0;
	pthread_mutex_unlock(&t);
  f(&t);
}
int main() {
  int i = 1; // not shared here
	pthread_t t1;

  pthread_create(&t1, NULL, dostuff, &i);
  pthread_create(&t1, NULL, dostuff, &i);
	if (i) {
		i = 0;
	}
}


