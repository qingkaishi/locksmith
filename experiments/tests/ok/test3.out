************************* STARTING *************************

functions declared and used but not defined:
  printf
shared: &i:test3.c:28

Warning: Possible data race: &i:test3.c:28 is not protected!
 references:
  dereference of cast:test3.c:23 at test3.c:23
    &i:test3.c:28 => cast:test3.c:23 at dostuff test3.c:31
  locks acquired:
    concrete *l::-1 -> dostuff test3.c:31 -> f test3.c:25
  in: main at test3.c:27 

  dereference of cast:test3.c:23 at test3.c:23
    &i:test3.c:28 => cast:test3.c:23 at dostuff test3.c:31
  locks acquired:
    concrete *l::-1 -> dostuff test3.c:32 -> f test3.c:25
  in: main at test3.c:27 

  dereference of &i:test3.c:28 at test3.c:33
    &i:test3.c:28
  locks acquired:
    <empty>
  in: main at test3.c:27 

  dereference of &i:test3.c:28 at test3.c:34
    &i:test3.c:28
  locks acquired:
    <empty>
  in: main at test3.c:27 


*************************** DONE ***************************
